### ISSUE TITLE
Integrating easy-profiler for light weight profiling

### ISSUE THREAD

**Issue by kjetilly** [OPM/opm-common#3248]
Adding this issue to kick off the discussion of integrating easy_profiler into OPM.

We frequently need to time certain sections of code, and in OPM, the standard way of doing this is typically to dump the timing output to a log and then parse (or manually read) the timings from there. This process is both needlessly time-consuming and prone to errors. 

While using a dedicated profiler such as `prof`, `callgrind` or the Intel Suite is less intrusive for the code itself, it often comes with a more involved setup phase, and fine-grained timing of inlined code (which appears frequently in OPM) or of specific sections of code, can be impossible with this approach. 

Hence we suggest integrating [easy_profiler](https://github.com/yse/easy_profiler) into OPM. easy_profiler lets you time specific sections of code by inserting macro calls into the code. eg:

```C++
EASY_BLOCK("heavy block");
for(size_t i = 0; i < (1<<42); ++i) { 
  // compute everything
}
```

By integrating easy_profiler, you can easily get timings on specific sections of code without having to parse through log files, and it comes with a GUI (optional) for analyzing the data. You can also watch profiling output live on your computer or a remote machine. 

It's a rather light weight option, and we've used it in other projects and have had good experiences with it. We would suggest adding this to OPM as an _optional build depedency_. Concretetly, we would suggest adding the following wrapper macros

```C++
// in say <opm/common/easy_timer.hpp>
#if OPM_HAVE_EASY_PROFILER
#include <easy/profiler.h>
#define OPM_TIME_BLOCK(blockname) EASY_BLOCK(blockname)
// other easy_profiler macros
#else
#define OPM_TIME_BLOCK(blockname)
// other easy_profiler macros
#endif
```

This way one can without issue build and run OPM without easy_profiler, and the code will not be majorly affected by it. To time specifc pieces of code, one would then insert macro calls of the form

```C++
OPM_TIME_BLOCK("blockname");
```

**Comment by akva2** [comment]
This makes a lot of sense. It will allow us to add profiling sections once and for all instead of developers keep adding their own local hacks whenever something has to be profiled. It's non-intrusive as the macros will carry no code unless you build against the library, and nobody excepts developers will ever need to build the dependency. Not being packaged is thus not important, as whoever wants to use this will also be capable of building their own. I also noticed the lack of recent commits, but that can also be seen as a positive; it does its job, is relatively bug free and users are happy with its functionality. While prof and callgrind is useful for looking at cache misses etc, they are very awkward (imo) to use for timing sections of code. TLDR; I'm positive to this.

**Comment by atgeirr** [comment]
I have usually used Instruments (a sampling-based non-intrusive profiler) and been quite happy with it, but I see the need for more detailed and specific information, some of which cannot be reliably collected by sampling (i.e. call counts).

My main worry is the readability of the code, if we add 100s of such blocks. I think it will be fine though, with some discipline the blocks can be a form of documentation. For example, do not use "heavy block" but "interpolate in 2D table" or more specific block tags, and it will be fine.

We can choose to be a bit restrictive about how many such block instances we put in the code if this becomes messy, developers can go crazy with them on test branches if they want. I think we can use this for much quicker debugging of performance issues if we keep a good set of them in the main branch, so I'd suggest we start with a "Yes" policy though.

**Comment by alfbr** [comment]
Developers do care whether third party dependencies are packaged or not. Dune, libecl, superlu, umfpack, zoltan comes to mind.

Yes, external profilers are useful, but they do not fulfill the need for internal profiling. However, we already do quite extensive internal profiling of many blocks of code. The basic building block is Dune::Timer I believe. Hence, I am wondering what added benefit we see bringing in another framework? If it is significant, are we better off improving what we have to cater to the needs?

The lack of commits is not positive in this case I am afraid. The project is already subject to bit rot. It has been a one-man-show for the last three years. Moreover, that person left the building a long time ago and the project resides on that persons personal account on github. There is almost as many open issues as closed issues. Pull requests stay open for years without response. Hence, if we are to take in this code, we should copy it into our repositories and realize that we need to take the investment of maintaining it.

**Comment by blattms** [comment]
To be honest, I do not know anything about this software. If this becomes a dependency, then this has to behave like a no-op if it is not found. But I guess this is the intention anyway,

One of the reasons for this is that (as @alfbr  pointed out) the repo seems dead since June 2021. As it seems to be C++ (there is a comment on depending on a C++11 compiler), this might mean that it becomes unusable in a few years. Just something that people need to be aware of before investing time.

Since you guys have tried it:
- How is the timing done exactly? Seems like this is done in os-specific way.
- How does it behave in a parallel program?

**Comment by alfbr** [comment]
Please do not interpret my comments as negative, I am just trying to explore the implications. I do see that having internal profiling that can be turned on and off at compile time is interesting. Right now, all internal profiling is done, and there is no way to turn it off (however, this small feature may be easy to implement with what we already have). I am probably one of the people who have used those profiling numbers the most, so improvements here is definitely of interest. We can consider easy_profiler, but then we should also consider replacing the current code blocks timings with easy_profiler. Having multiple frameworks doing the same thing is not efficient use of scarce developer resources. We should also be clear that then this is what we use (how else would we convince people to use this instead of hacks?). However, the easy_profiler project is dead, and the code may disappear at any point in time, so we probably need to take it over wholesale if we start using it in master. Whatever people use in their own branches and forks are of course entirely up to themselves.

**Comment by kjetilly** [comment]
With regards to the inactivity of the repository: This is a valid point. However, easy-profiler will not be tightly integrated into OPM. With the macro we suggested, one could easily change to alternative solutions such as [tracy](https://github.com/wolfpld/tracy), [palanteer](https://github.com/dfeneyrou/palanteer), [microprofile](https://github.com/jonasmr/microprofile) or [Remotery](https://github.com/Celtoys/Remotery) or a simple `Dune::Timer` fallback by just changing the macro definition (some of these require a setup step, so one would have a setup function that could change as well).

While the above mentioned libraries all are more active than easy-profile, they are in our experience slightly more cumbersome to start using. We are currently testing out tracy as well, and while it does look a lot nicer and have sampling profiling as an added benefit, it requires you to run the tracy server whenever you want to capture performance data. 

As far as dependency management goes, I believe easy-profiler can be added through say either `FetchContent` or `ExternalProject` in cmake, in other words, no extra build step for the developer. We have successfully tested this with tracy (with the caveat that one still has to built the server with tracy) and it seemed to work well. (as an Arch user (btw) I will also mention that both easy-profiler and tracy **are** available in a package manager :) )

For a proof of concept, I would suggest making the macro agnostic to tracy, easy-profiler and `Dune::Timer`. Something like

```C++
#ifdef OPM_USE_EASY_PROFILER
    #include <easy/profiler.h>
    #define OPM_TIME_BLOCK(blockname) EasyBlock(#blockname)
#elif OPM_USE_TRACY
    #include <tracy/Tracy.hpp>
    #define OPM_TIME_BLOCK(blockname) ZoneNamedN(blockname, #blockname, true)
#elif OPM_USE_DUNE_TIMER
    #include <relevant wrapper header>
    #define OPM_TIME_BLOCK(blockname) //Some wrapper around Dune::Timer that logs the time
#else
    #define OPM_TIME_BLOCK(blockname)
#endif
```

So for the technical questions:

> How is the timing done exactly? Seems like this is done in os-specific way.

I am not sure, but I can only assume they use a timer that does not involve a context switch. This is one of the reasons using a library for this is a good idea, since getting microbenchmarking done correctly can be very hard.

> How does it behave in a parallel program?

For threads it works fine, for MPI easy-profiler works and gives you the result per process. For tracy, MPI needs some special care if you want output from each process and not just the one process.

**Comment by blattms** [comment]
> We can consider easy_profiler, but then we should also consider replacing the current code blocks timings with easy_profiler.

That sounds like making it a required dependency. At least if you also mean the timings meant that get printed in a default production run.

**Comment by kjetilly** [comment]
I would also not recommend using easy-profiler or alternatives for timings that get printed in a default production run. At least not at the default verbosity level. Now, there might be some finer granularity timings happening at larger verbosity levels in there that could be gradually phased out with easy-profiler or alternatives instead. I am not even sure if easy-profiler supports getting the timing data during program execution from the program itself. While it's probably technically possible, it does not seem like a feature they advertise.

**Comment by alfbr** [comment]
> > We can consider easy_profiler, but then we should also consider replacing the current code blocks timings with easy_profiler.
> 
> That sounds like making it a required dependency. At least if you also mean the timings meant that get printed in a default production run.

Yes, but that is based on my assumption that we should unify the internal timers rather than having multiple co-existing solutions. Which again is based on my assumption that the code base is confusing/complex enough already.

> As far as dependency management goes, I believe easy-profiler can be added through say either FetchContent or ExternalProject in cmake

That is basically what I am saying is probably not a viable option. The code may disappear from that github account at any point in time. If we start investing in it, I am in favour of forking it into our repositories (or possibly the parts of it we want). That will give us full flexibility while making it clear that we need to maintain it ourselves. This exemplifies why taking in third party libraries should not be done lightly. Once we start investing in it (in master, not considering forks or branches), we quickly become dependent on it. Having it optional in the build system is fooling ourselves in this respect. Flow is not a hobby project at this point, it is production code. After all, the main argument at the start was avoiding hacks, which entails multiple developers investing and collaborating on it.

**Comment by kjetilly** [comment]
I absolutely see the concern about maintainability and dependency of an external project.

If the decision lands on easy-profiler (after thorough testing across platforms, of course), I would suggest the following:

1) Fork easy-profiler to github.com/OPM
2) Point cmake {ExternalProject, FetchContent} to github.com/OPM/easy-profiler

Then if we run into issues with easy-profiler down the line (say for C++ compiler compatibility issues), I would rather suggest jumping ship to an alternative (tracy, palanteer, microprofile, Remotery or whatever else is around at that time) rather than trying to fix easy-profiler. 

In terms of code lines that need to change this would be a very small change (basically just change the macro and update the cmake files).

For archiving purposes of old timings, easy-profiler comes with the utility program `profiler_converter` which converts the files to a JSON formatted file that, while verbose, is not difficult to convert to a different format. tracy on the other hand lets you dump the result to a CSV file (which is slightly less verbose)

All that being said, I am not dead fixed on easy-profiler, but among the alternatives, it strikes me as the easiest option to get up and running. We will test the other options as well, of course.

**Comment by blattms** [comment]
@alfbr Just be aware that if this will become a necessary dependency, then this will be increase the packaging effort and we are already quite time constrained...

**Comment by alfbr** [comment]
> @alfbr Just be aware that if this will become a necessary dependency, then this will be increase the packaging effort and we are already quite time constrained...

Yes, but this seems not to be on the table at this point in time. As a first step it looks like it will come in addition to what we already have, as easy_compiler is not really set up to integrate into our reporting (i.e., logging to one of our output files during simulations). Hence, optional dependency only for those interested in additional profiling. I still think it would make sense to unify our internal profiling, but I guess it does not hurt to get some experience from alternatives before going down that road anyway. It would be great if more developers could give their thoughts though.

**Comment by kjetilly** [comment]
I did a very short [proof of concept of integrating different benchmarking utilities with one macro](https://github.com/kjetilly/benchmark_testing). 

We have one standard macro `PROGRAM_TIME_BLOCK(blockname)` that is used in the code. For example, see [main.cpp](https://github.com/kjetilly/benchmark_testing/blob/main/src/main.cpp) for the usage of this macro.

Based on CMake options, the correct macro backend is then decided at compile time seamlessly. So specifying say

```bash
cmake -DPROGRAM_TIMER_BACKEND=TRACY
```
uses Tracy for profiling, while

```bash
cmake -DPROGRAM_TIMER_BACKEND=EASY_PROFILER
``` 
uses easy_profiler. Also supported are palanteer, remotery and microprofile. 

Now, this is not the prettiest code I've ever written (especially the cmake stuff), but it does show that switching between different profiling backends can be done quite easily. There are a ton of design decisions that can be argued this way and that as well, but those I claim would be minor details. 

Additionally, I have written some very simple cmake `FetchContent` snippets to fetch the relevant profiling libraries on demand with no intervention from the developer. Meaning the developer does not need to download the profiler libraries separately (they do, however, need to install the clients/viewers somehow, though that can be done through say a docker container or even a flatpak). If we do go this route, we should fork the repos we fetch for maintainability and security. For now, I've opted to download the libraries from the repositories directly since it's a proof of concept.

Now, some of the libraries (easy profiler, tracy and more) support custom colors for the blocks. This is also some we could support down the line by defining additional color macros that translate to the relvant ones when the library supports it, and to nothing if the library does not support colors.
